Implement autodetection for RISC-V

leekillough opened the pull request:
Implement autodetection for RISC-V, whose architecture can be detected with preprocessor macros alone. This avoids having to run RISC-V binaries on a cross-compiler host.

leekillough committed:
Implement autodetection for RISC-V, whose architecture can be detected with preprocessor macros alone. This avoids having to run RISC-V binaries on a cross-compiler host.

angsch commented on frame/base/bli_riscv_cpuid.h:
```diff
@@ -0,0 +1,15 @@
+#if __riscv_xlen >= 64
```
`if __riscv_xlen == 64`

angsch commented on frame/base/bli_riscv_cpuid.h:
```diff
@@ -0,0 +1,15 @@
+#else
+rv64i
+#endif
+#elif __riscv_xlen >= 32
```
`#elif __riscv_xlen == 32`

angsch reviewed (commented):
`generic` is currently unreachable -- 16-bit mode does not exist in RISC-V. With the proposed fix `generic` will be picked up with 128-bit cores. I think that this is at the very least more correct. Recall the issues with loading and storing of callee-saved registers with rv32 and rv64.

However, there is one more _potential_ problem: Is is correct to assume that `I` is available throughout all platforms? Shouldn't `generic` be picked up if none of the hardware requirements are satisfied? 

Please add the BLIS copyright header.

leekillough commented:
> `generic` is currently unreachable -- 16-bit mode does not exist in RISC-V. With the proposed fix `generic` will be picked up with 128-bit cores. I think that this is at the very least more correct. Recall the issues with loading and storing of callee-saved registers with rv32 and rv64.

I have added comments to the code to make it clearer.

I have tweaked it so that 128-bit mode gets detected as `generic`, and have fixed the assembly code macros for callee-saved registers to support 128-bit, to make them future-proof and available for new code, while not supporting 128-bit just yet.

> Not all compilers support the architecture extension test macros. I managed to build a compiler with rvv that does not support __riscv_vector. I expect this problem to disappear with time.

`rvv-next` [seems not to be maintained](https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1177#issuecomment-1383156444).  If a toolchain (not just a compiler, but headers too) does not support the architecture test macros, then someone will need to manually specify `rv32i`, `rv32iv`, `rv64i` or `rv64iv` with BLIS `configure`. That does not seem like a huge burden.

> Please add the BLIS copyright header.

Done.

angsch commented:
>  I don't understand. 16-bit mode is not tested for, nor was I aware of an int16 (short) RISC-V spec. In any case, if __riscv_xlen == 16, then the tests will fail, and we will fall back on generic.

You misread. 16-bit mode does *not* exist. 

rv128iv will go into rv64i. It defines the v extension, and then goes into the else branch. This would crash. As stated by mail, I don't have a strong opinion on the order of targets, but I do consider it a necessity to check xlen with `==` and not `>=`. As a corner case, rv32ev should likely go into generic and not, as it is the case now, rv32iv. This is where the proposal to rearrange the if-else construct comes from: Top-level covers extensions, inner level checks for xlen.

leekillough committed:
Add __riscv_i macro check to exclude embedded reduced ISAs

Fall back to generic on 128-bit RISC-V

Add support for saving/restoring 128-bit callee-saved registers
(making code future-proof but not supporting 128-bit right now);
change the saving/restoring code to only use as much stack as
necessary

Add comments and copyright to the RISC-V autodetection code

leekillough commented:
> You misread. 16-bit mode does _not_ exist.

I was confused by the wording. We often speak in bits, bytes, or vector words, and get them confused.

I have added `__riscv_i` checks, and made the code fall back to `generic` if `__riscv_i` isn't defined or if `__riscv_xlen == 128` .

Only when `__riscv_i && __riscv_xlen == 64` is `rv64i` or `rv64iv` selected, and only when `__riscv_i && __riscv_xlen == 32` is `rv32i` or `rv32iv` selected.

I squashed a few commits, but not the original commit of this PR that you previously reviewed (comments edited above).

angsch reviewed (approved):
Thank you.
